movieland-service-impl/src/main/java/com/gromoks/movieland/service/impl/UserCacheImpl.java

```diff
@@ -0,0 +1,120 @@
+package com.gromoks.movieland.service.impl;
+
+import com.gromoks.movieland.dao.jdbc.UserDao;
+import com.gromoks.movieland.entity.User;
+import com.gromoks.movieland.service.UserCache;
+import com.gromoks.movieland.service.entity.LoginRequest;
+import com.gromoks.movieland.service.entity.UserToken;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+public class UserCacheImpl implements UserCache{
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    private final Map<String,UserToken> userTokenMap = new ConcurrentHashMap<>();
+
+    @Value("${cache.user.token.timeToLiveInMs}")
+    private long timeToLiveInMs;
+
+    @Autowired
+    private UserDao userDao;
+
+    public UserCacheImpl() {
+
+    }
+
+    public UserCacheImpl(UserDao userDao) {
+        this.userDao = userDao;
+    }
+
+    @Override
+    public UserToken getUserToken(LoginRequest loginRequest) {
+        if (userTokenMap.containsKey(loginRequest.getEmail())) {
+            validateLoginRequestParameter(loginRequest);
+            return userTokenMap.get(loginRequest.getEmail());
+        } else {
+            return  generateUserToken(loginRequest);
+        }
+    }
+
+    @Override
+    public void removeUserToken(String uuid) {
+        log.info("Start to remove user token with uuid = {}",uuid);
+        for (String key : userTokenMap.keySet()) {
+            if (userTokenMap.get(key).getUuid().equals(uuid)) {
+                userTokenMap.remove(key);
+                log.info("Removed user token");
+                break;
+            }
+        }
+    }
+
+    @Override
+    @PostConstruct
+    @Scheduled(fixedRateString="${cache.fixedRate.user.token}")
+    public void invalidate() {
+        log.info("Invalidate expired uuid");
+        for (String key : userTokenMap.keySet()) {
+            if (System.currentTimeMillis() - userTokenMap.get(key).getInitTimeInMs() >= timeToLiveInMs) {
```

Comment:
to avoid calculation there, you can save time of token end, but not time of token init